V3 background sync

philipwalton opened the pull request:
R: @jeffposnick @addyosmani @gauntface

Changes described in #868.

jeffposnick commented on packages/workbox-background-sync/lib/StorableRequest.mjs:
```diff
@@ -0,0 +1,116 @@
+   * new Request object. A timestamp is also set in the event that when this
+   * object was created is relevant after it's stored.
+   *
+   * @param {StorableRequest|Object} param1
```
I don't follow this JSDoc—what's `{StorableRequest|Object}`?

philipwalton replied:
It means it could be of either type. In this case a `StorableRequest` object could be created from a plain object or "cloned" from another `StorableRequest` instance.

However, looking at the code, I think I only create this from plain objects, so I can remove.

jeffposnick commented on packages/workbox-background-sync/lib/StorableRequest.mjs:
```diff
@@ -0,0 +1,116 @@
+   *     defaulting to the current time if not specified.
+   */
+  constructor({url, requestInit, timestamp = Date.now()}) {
+    this.url = url;
```
Matt's been pretty keen on prefixing all instance variable names with `_` as that apparently helps with our uglification.

gauntface replied:
👍 Unless you expect people to use `instance.url` but even then, in some situations getters and setters can be smaller when using the private variable lots inside the class.

philipwalton replied:
This needs to be unprefixed as it's public and can/must be editable by the user. E.g. `workbox-google-analytics` will need to modify these values before requests are replayed.

gauntface commented on infra/testing/sw-env.js:
```diff
@@ -11,21 +11,158 @@
+const mockFetch = require('./mock-fetch');
+
+Object.assign(global, makeServiceWorkerEnv());
+global.self = global;
```
Why are you assigning global to `global.self`? makeServiceWorkerEnv has a `self` property which you'll now we overriding with everything node has.

philipwalton replied:
This was necessary to handle the mocks we're adding later. I could try doing the global assignment after adding the mocks to see if that works. I remember there being an issue that this solve, but I can't remember all the various things I tried.

gauntface commented on infra/testing/sw-env.js:
```diff
@@ -11,21 +11,158 @@
-const swEnv = makeServiceWorkerEnv();
+// Stub missing/broken Headers API methods in `service-worker-mock`.
+// https://fetch.spec.whatwg.org/#headers-class
+class Headers {
```
Please move all the mocks into a seperate file to clean up this file.

philipwalton replied:
BTW, the end goal was to PR this back into `service-worker-mock`.

I can move it, but regardless this was always supposed to be a temporary solution.

philipwalton replied:
Done.

gauntface commented on infra/testing/sw-env.js:
```diff
@@ -11,21 +11,158 @@
+  }
+};
+
+global.__removeAllEventListeners = () => {
```
nit: rename to __resetAllEventListeners. Would also be good if you move this to a seperate file and rather than expose this method on global, expose a function from the new file can all tests to call it (This is what the indexedDB module does - feels a little cleaner).

philipwalton replied:
Agreed, though similar to my above comment, I was going to feature request this for the original library. We can discuss whether it makes sense to spend time on this now or spend time on a PR.

philipwalton replied:
Done.

philipwalton replied:
Done.

gauntface commented on infra/testing/sw-env.js:
```diff
@@ -11,21 +11,158 @@
-swEnv.location = 'https://example.com';
+global.location = 'https://example.com';
 
+// Needed to look like a real SW env
```
Just an FYI - when you rebase, these next few lines will be removed (due to fix in service-worker-mock.

philipwalton replied:
Cool, rebased.

gauntface commented on packages/workbox-background-sync/lib/Queue.mjs:
```diff
@@ -0,0 +1,297 @@
+
+
+const {indexedDBHelper, WorkboxError} = _private;
+const names = new Set();
```
This feels like an unusual pattern - I think you are doing it to ensure a common set of names across all instances of Queue. Would be good to add a comment

philipwalton replied:
What do you mean? Using a set rather than an array?

gauntface commented on packages/workbox-background-sync/lib/Queue.mjs:
```diff
@@ -0,0 +1,297 @@
+ * later. All parts of the storing and replaying process are observable via
+ * callbacks.
+ *
+ * @example
```
👍

philipwalton replied:
Done.

gauntface commented on packages/workbox-background-sync/lib/Queue.mjs:
```diff
@@ -0,0 +1,297 @@
+   *     unique as it's used to register sync events and store requests
+   *     in IndexedDB specific to this instance. An error will be thrown if
+   *     a duplicate name is detected.
+   * @param {Object} [param2]
```
👍

philipwalton replied:
This is a convention [mentioned in the Google JS style guide](https://google.github.io/styleguide/jsguide.html#features-arrays-destructuring).

I suppose we can discuss what convention we want to follow, but the point is to support cases where you could have multiple destructured parameters in the same function signature.

gauntface replied:
We aren't saying that you can't destructure. Just give a name to the  object for jsdocs.

gauntface commented on packages/workbox-background-sync/lib/Queue.mjs:
```diff
@@ -0,0 +1,297 @@
+   * @param {Object} [param2]
+   * @param {number} [param2.maxRetentionTime = 7 days] The amount of time (in
+   *     ms) a request may be retried. After this amount of time has passed,
+   *     the request will be deleted and not retried.
```
Nit .: reword to "deleted from the queue." insteand of "deleted and not retried."

philipwalton replied:
Done.

gauntface commented on packages/workbox-background-sync/lib/Queue.mjs:
```diff
@@ -0,0 +1,297 @@
+    const storableRequests = [];
+
+    for (const [key, entry] of (await db.getAll()).entries()) {
+      if (entry.queueName == this._name) {
```
=== please

gauntface commented on packages/workbox-background-sync/lib/Queue.mjs:
```diff
@@ -0,0 +1,297 @@
+    const db = await this._getDb();
+    const storableRequests = [];
+
+    for (const [key, entry] of (await db.getAll()).entries()) {
```
Could you pull out the await piece. This is a fairly gnarly looking chunk of code.

gauntface commented on packages/workbox-background-sync/lib/Queue.mjs:
```diff
@@ -0,0 +1,297 @@
+
+    for (const [key, entry] of (await db.getAll()).entries()) {
+      if (entry.queueName == this._name) {
+        // Requests older than `maxRetentionTime` should be ignored.
```
Nit: change to "should be removed."

gauntface commented on packages/workbox-background-sync/lib/Queue.mjs:
```diff
@@ -0,0 +1,297 @@
+
+        if (Date.now() - storableRequest.timestamp > this._maxRetentionTime) {
+          // No need to await this since it can happen in parallel.
+          this._removeRequest(key);
```
If all storage requests are expired, will anything waitUntil these operations have completed or are we just hoping it works?

gauntface commented on packages/workbox-background-sync/lib/Queue.mjs:
```diff
@@ -0,0 +1,297 @@
+          continue;
+        }
+
+        storableRequests.push([key, storableRequest]);
```
You're creating an array of arrays. Would a map or object be better here?

gauntface commented on packages/workbox-background-sync/lib/Queue.mjs:
```diff
@@ -0,0 +1,297 @@
+  async _registerSync() {
+    try {
+      await this._waitUntilActive();
+      await registration.sync.register(`${TAG_PREFIX}:${this._name}`);
```
If the queue name is unique, why are we adding a prefix? We can just use the queue name. Would be what the developer probably expects.

gauntface commented on packages/workbox-background-sync/lib/StorableRequest.mjs:
```diff
@@ -0,0 +1,116 @@
+
+  /**
+   * Accepts a URL and RequestInit dictionary that can be used to create a
+   * new Request object. A timestamp is also set in the event that when this
```
Nit: can you re-word "A timestamp is also set in the event that when this object was created is relevant after it's stored." I'm not understanding what it means.

philipwalton replied:
Done.

gauntface commented on packages/workbox-background-sync/package.json:
```diff
@@ -0,0 +1,30 @@
+    "sync"
+  ],
+  "scripts": {
+    "prepublish": "gulp build-packages --package workbox-background-sync"
```
This should be changed to "build" and add "prepare" and "version" npm run scripts to call `npm run build` see modules in v3 branch.

philipwalton replied:
Done.

gauntface commented on test/workbox-background-sync/node/lib/test-StorableRequest.mjs:
```diff
@@ -0,0 +1,142 @@
+    '../../../../packages/workbox-background-sync/lib/StorableRequest.mjs';
+
+
+describe(`StorableRequest`, function() {
```
Please add `[workbox-background-sync]` to the top level describe call.

philipwalton replied:
Added, but I originally didn't add anything here since this object isn't publicly exposed. Do you have thoughts on how we annotate that?

gauntface commented on test/workbox-background-sync/node/lib/test-Queue.mjs:
```diff
@@ -0,0 +1,535 @@
+
+  beforeEach(async function() {
+    sandbox.restore();
+    clearRequire.match(RegExp('workbox-background-sync/lib/Queue.mjs'));
```
add a comment that this is to clear the names in the Queue.mjs.

philipwalton replied:
Done.

gauntface commented on test/workbox-background-sync/node/lib/test-Queue.mjs:
```diff
@@ -0,0 +1,535 @@
+    sandbox.restore();
+    clearRequire.match(RegExp('workbox-background-sync/lib/Queue.mjs'));
+
+    clearObjectStore();
```
the indexedDB mock has a reset method - see precaching tests.

philipwalton replied:
Unfortunately, that doesn't work. I tried it originally, but `reset` just clears the references, it doesn't delete the objects already in memory.

gauntface commented on test/workbox-background-sync/node/lib/test-Queue.mjs:
```diff
@@ -0,0 +1,535 @@
+    clearObjectStore();
+
+    // Remove any lingering event listeners
+    global.__removeAllEventListeners();
```
as mentioned above, would be nice if this followed the indexedDB mock model of exposing a function that could be imported and run to reset.

philipwalton replied:
Done.

gauntface commented on test/workbox-background-sync/node/lib/test-Queue.mjs:
```diff
@@ -0,0 +1,535 @@
+    it(`should try to replay the queue on SW startup in browsers that ` +
+        `don't support the sync event`, async function() {
+      // Delete the SyncManager interface to mock a non-supporting browser.
+      const originalSyncManager = registration.sync;
```
you *might* be able to do `sanbox.stub(registration, 'sync').value(undefined)` and that you don't need to track original or perform delete

philipwalton replied:
No, I tried that as well as `.value()` (leaving it empty) and neither worked. Probably a good feature request for sinon.

gauntface commented on test/workbox-background-sync/node/lib/test-Queue.mjs:
```diff
@@ -0,0 +1,535 @@
+
+      expect(Queue.prototype.replayRequests.calledOnce).to.be.true;
+
+      registration.sync = originalSyncManager;
```
If the expect fails and throws, the original will never be set. Might be best to store the original in the describe and re-set in beforeEach and after (if the sinon approach doesn't work)

philipwalton replied:
If the tests fail, do we care if state gets messed up? I could also do something like this:

```js
try {
  expect(Queue.prototype.replayRequests.calledOnce).to.be.true;
} catch (err) {
  throw err;
} finally {
  registration.sync = originalSyncManager;
}
```

gauntface commented on test/workbox-background-sync/node/lib/test-Queue.mjs:
```diff
@@ -0,0 +1,535 @@
+  });
+
+  describe(`createPlugin`, function() {
+    it(`should return an object implementing the fetchDidFail plugin ` +
```
This method still feels weird to me. May be naming, but I'm struggling to understand what it's for.

gauntface commented on test/workbox-background-sync/node/lib/test-Queue.mjs:
```diff
@@ -0,0 +1,535 @@
+      await queue.addRequest(request);
+
+      // Allow time to ensure async registration didn't happen.
+      await sleep(100);
```
This feels flakey and unnecessary. This doesn't seem to affect the flow at all of this test. If you do need timeouts please use sinon fake timers.

gauntface commented on test/workbox-background-sync/node/lib/test-Queue.mjs:
```diff
@@ -0,0 +1,535 @@
+    it(`should ignore (and remove) requests if maxRetentionTime has passed`,
+        async function() {
+      sandbox.spy(self, 'fetch');
+      const clock = sinon.useFakeTimers({
```
Can you get this from the sandbox? If so you can drop the restore call at the end of this test.

philipwalton replied:
Wasn't in the documentation, but yep, appears to work. Done.

philipwalton commented on infra/testing/sw-env.js:
```diff
@@ -11,21 +11,158 @@
+
+    this.url = url;
+    this.method = options.method || 'GET';
+    this.mode = options.mode || 'same-origin'; // FF defaults to cors
```
Sure, I'll do that.

philipwalton commented on infra/testing/sw-env.js:
```diff
@@ -11,21 +11,158 @@
+    this.mode = options.mode || 'same-origin'; // FF defaults to cors
+    this.headers = new Headers(options.headers);
+
+    this.bodyUsed = !!options.body;
```
Done.

philipwalton commented on infra/testing/sw-env.js:
```diff
@@ -11,21 +11,158 @@
+    }
+  }
+
+  clone() {
```
Done.

philipwalton commented on infra/testing/sw-env.js:
```diff
@@ -11,21 +11,158 @@
+    return new Request(this.url, this);
+  }
+
+  text() {
```
Done.

philipwalton commented on packages/workbox-background-sync/lib/Queue.mjs:
```diff
@@ -0,0 +1,297 @@
+    const storableRequestsInQueue = await this._getStorableRequestsInQueue();
+
+    // If nothing is in the queue, return immediately and run no callbacks.
+    if (!storableRequestsInQueue.length) return;
```
This is allowed per Google style. Short, concise, one-line `if` statements don't need braces.

philipwalton commented on packages/workbox-background-sync/lib/Queue.mjs:
```diff
@@ -0,0 +1,297 @@
+    for (const [key, entry] of (await db.getAll()).entries()) {
+      if (entry.queueName == this._name) {
+        // Requests older than `maxRetentionTime` should be ignored.
+        const storableRequest = new StorableRequest(entry.storableRequest);
```
Done.

philipwalton commented on packages/workbox-background-sync/lib/Queue.mjs:
```diff
@@ -0,0 +1,297 @@
+        // Requests older than `maxRetentionTime` should be ignored.
+        const storableRequest = new StorableRequest(entry.storableRequest);
+
+        if (Date.now() - storableRequest.timestamp > this._maxRetentionTime) {
```
Done.

philipwalton commented on packages/workbox-background-sync/lib/Queue.mjs:
```diff
@@ -0,0 +1,297 @@
+   *
+   * @private
+   * @param {string} name The name of the callback on this._callbacks.
+   * @param {...*} args The arguments to invoke the callback with.
```
Yep, you can see examples [here](https://google.github.io/styleguide/jsguide.html#features-functions-rest-parameters) and [here](https://github.com/google/closure-compiler/wiki/Annotating-JavaScript-for-the-Closure-Compiler#variable-parameters-in-param-annotations).

gauntface replied:
Can you confirm this works in JSDocs with `gulp docs`

philipwalton replied:
Oh cool, I wasn't aware we could do this. Upon initial inspection there's lots of stuff that's missing, so I'll do a cleanup and resubmit.

philipwalton replied:
Done.

philipwalton commented on packages/workbox-background-sync/lib/Queue.mjs:
```diff
@@ -0,0 +1,297 @@
+   * @private
+   */
+  _addSyncListener() {
+    self.addEventListener('sync', (event) => {
```
Done.

philipwalton commented on packages/workbox-background-sync/lib/Queue.mjs:
```diff
@@ -0,0 +1,297 @@
+   * @private
+   * @return {Promise}
+   */
+  _waitUntilActive() {
```
Nope, happy to remove the logic if it's not needed.

It looks like maybe the throwing cases is primarily to support registration from the window context?

philipwalton commented on packages/workbox-background-sync/lib/StorableRequest.mjs:
```diff
@@ -0,0 +1,116 @@
+ limitations under the License.
+*/
+
+const serializableProperties = [
```
I haven't confirmed, but I only include these in the `requestInit` dict if they're present.

But, either way, yeah, it probably doesn't make sense to include properties that are only handled by the UA. I'll look into this further.

philipwalton commented on packages/workbox-background-sync/lib/StorableRequest.mjs:
```diff
@@ -0,0 +1,116 @@
+    const requestInit = {headers: {}};
+
+    // Set the body, if used.
+    if (request.bodyUsed) {
```
Yeah, this was in the original code, but I see now that it's not what I need.

Given this, do you have a recommendation for how to best get the body content? Ideally in a way that can support https://github.com/GoogleChrome/workbox/issues/863.

philipwalton commented on packages/workbox-background-sync/lib/Queue.mjs:
```diff
@@ -0,0 +1,297 @@
+  /**
+   * Retrieves all stored requests in IndexedDB and retries them. If the
+   * queue contained requests that were successfully replayed, the
+   * `allRequestsDidReplay` callback is invoked (which implies the queue is
```
`allRequestsDidReplay` is invoked with an array of `{request, response}` objects, so the user could inspect the array to see if there were any 4XX responses and then deal with them however they want.

Do you think this is enough?

philipwalton commented on packages/workbox-background-sync/lib/Queue.mjs:
```diff
@@ -0,0 +1,297 @@
+   * now empty). If any of the requests fail, a new sync registration is
+   * created to retry again later.
+   */
+  async replayRequests() {
```
It is more like a controller. I could abstract some of the queue-ing logic into a separate module, but I'm honestly not sure how much cleaner that would be.

The only methods where it'd made sense to do that are `_getStorableRequestsInQueue()`, `_getDB()`, and `_removeRequest()`.

philipwalton commented on packages/workbox-background-sync/lib/Queue.mjs:
```diff
@@ -0,0 +1,297 @@
+    const storableRequestsInQueue = await this._getStorableRequestsInQueue();
+
+    // If nothing is in the queue, return immediately and run no callbacks.
+    if (!storableRequestsInQueue.length) return;
```
Yeah, I think that makes sense. Updated.

philipwalton commented on packages/workbox-background-sync/lib/Queue.mjs:
```diff
@@ -0,0 +1,297 @@
+        continue;
+      }
+
+      this._runCallback('requestDidReplay', {request, response});
```
Good call! Done.

philipwalton commented on packages/workbox-background-sync/lib/Queue.mjs:
```diff
@@ -0,0 +1,297 @@
+      // Remove the request from IndexedDB asynchronously (don't await).
+      // TODO(philipwalton): in the unlikely event that the delete fails,
+      // this request may be replayed again. Do we want to handle this case?
+      this._removeRequest(key);
```
I went back and forth on this. The question is what would a developer do if the delete failed? They could retry, but then we'd have to expose a method to allow them to retry manually, and that seems not great.

An alternative approach I considered would be to remove items from the DB and then re-add them if the replay fails. The downside of that is it requires more DB reads/writes. The upside is if there's a read/write failure, a request is lost rather than replayed twice.

philipwalton commented on packages/workbox-background-sync/lib/Queue.mjs:
```diff
@@ -0,0 +1,297 @@
+   *
+   * @return {Object}
+   */
+  createPlugin() {
```
I added this because it was in the original API design. Here's the original code example (modified to use `createPlugin()`):

```js
const queue = new workbox.backgroundSync.Queue('myQueue');

const requestWrapper = new workbox.runtimeCaching.RequestWrapper({
  plugins: [queue.createPlugin()],
});

const route = new workbox.routing.RegExpRoute({
  regExp: new RegExp('^https://jsonplaceholder.typicode.com'),
  handler: new workbox.runtimeCaching.NetworkOnly({requestWrapper}),
});

const router = new workbox.routing.Router();
router.registerRoute({route});
```

gauntface replied:
I'd drop this in favor of exposing a plugin class like the other plugins. While I understand the convenience of this, nothing else follows this pattern.

philipwalton replied:
I don't like the v2 way (where the plugin class just extends the `Queue` class), but if you want it to be a class instead of an object, it could be like this:

```js
class QueuePlugin {
  constructor(queue) {
    this._queue = queue
  }
  fetchDidFail({request}) {
    this._queue.addRequest(request);
  }
}
```

And you'd use it like this:

```js
new workbox.runtimeCaching.NetworkOnly({
  plugins: new QueuePlugin(queue),
});
```

vs. how I have it currently:

```js
new workbox.runtimeCaching.NetworkOnly({
  plugins: queue.createPlugin(),
});
```

* * *

Alternatively, we could just do away with this feature entirely, since it's really not that hard to do this:

```js
new workbox.runtimeCaching.NetworkOnly({
  plugins: {
    fetchDidFail: ({request}) {
      queue.addRequest(request);
    }
  }
});
```

WDYT?

jeffposnick replied:
I'd vote for the first option.

If we go with explicitly needing to know the lifecycle callback names to use this behavior, then the precedent would be to use that approach with all the other plugins as well.

That isn't a great user experience for developers who care about specifying behaviors, but don't care about the underlying lifecycle callbacks. It's also an issue for the other behaviors that require multiple callbacks (I think cache expiration does?), since it's even harder to properly configure that.

philipwalton replied:
SGTM. Done.

philipwalton commented on packages/workbox-background-sync/lib/Queue.mjs:
```diff
@@ -0,0 +1,297 @@
+    }
+
+    if (allReplaysSuccessful) {
+      this._runCallback('allRequestsDidReplay', successfullyReplayedRequests);
```
I guess I could. Though if I did that I'm not sure why anyone would ever use the `replayDidFail` or `requestDidReplay` callbacks instead of this.

Maybe make this change and remove those two?

philipwalton commented on packages/workbox-background-sync/lib/Queue.mjs:
```diff
@@ -0,0 +1,297 @@
+    if (allReplaysSuccessful) {
+      this._runCallback('allRequestsDidReplay', successfullyReplayedRequests);
+    } else {
+      this._registerSync();
```
I guess not. I'll await this and the other occurrence of `_registerSync()`.

philipwalton commented on packages/workbox-background-sync/lib/Queue.mjs:
```diff
@@ -0,0 +1,297 @@
+   * @param {...*} args The arguments to invoke the callback with.
+   */
+  _runCallback(name, ...args) {
+    if (typeof this._callbacks[name] == 'function') {
```
Done. But what about the bytes Matt! The BYTES!

philipwalton commented on packages/workbox-background-sync/lib/Queue.mjs:
```diff
@@ -0,0 +1,297 @@
+   *
+   * @private
+   */
+  async _registerSync() {
```
@jeffposnick pointed out in another comment that I don't need to wait for activation here, so I've removed that part.

gauntface replied:
I'm assuming calling sync with the same details more than once get's de-duped by the browser?

philipwalton replied:
Yeah, that's in the spec.

philipwalton commented on packages/workbox-background-sync/lib/Queue.mjs:
```diff
@@ -0,0 +1,297 @@
+   *
+   * @private
+   */
+  _addSyncListener() {
```
I'm not sure I understand. Can you give more context?

Looking at those examples, they both expose a single instance of a class, whereas background sync exposes the class itself, and it needs a unique listener per instance.

jeffposnick replied:
https://github.com/GoogleChrome/workbox/blob/v3/packages/workbox-routing/default-export.mjs#L22 is an example of what I did for the `Router` class.

I had trouble wrapping my head around it as well, but once you get a feel for how the `default` exports are likely be used in the context of `workbox-sw`, it makes more sense.

philipwalton replied:
I still don't think this applies to the background sync case since I don't instantiate anything in the default export.

jeffposnick replied:
That should be okay by me, let's see if @gauntface has any concerns. He's given this setup a lot more thought that I have.

gauntface replied:
I still have concerns but this module is also too big to cover off everything. I'll play around with it and see if it makes to refactor after.

philipwalton replied:
Ok, I'd be happy to look at a suggestion for how it could be refactored, but since I don't understand what you're asking me to do, I can't really do it myself.

philipwalton commented:
@gauntface, @jeffposnick PTAL.

Changes since your last reviews:

- Per Matt's suggestion: added a `QueueStore` module to break up `Queue` logic.
- Per Matt's suggestion: removed the `requestDidFail` and `requestDidReplay` callbacks and changed `allRequestsDidReplay` to `queueDidReply` but now it runs anytime the queue is retried (previously it was only on success)
- Added support for Blobs in requests
- Removed the use of `indexedDBHelper` since I decided to use an index instead of `getAll()`.
- Per Matt's suggestion: moved sw-env mocks into their own files
- Per Matt's suggestion: removed the global `__removeAllEventListeners` in favor of exporting a `resetEventListeners` function from the event listener mock.

jeffposnick commented on packages/workbox-background-sync/lib/StorableRequest.mjs:
```diff
@@ -0,0 +1,119 @@
+    const requestInit = {headers: {}};
+
+    // Set the body if present.
+    if (request.method != 'GET') {
```
Could you switch here and elsewhere to `!==`? (Shouldn't our linting pick that up?)

philipwalton replied:
I believe the linting only warns against double-equals comparison with falsy values (as that's the only time it gets tricky).

I can change here, but I think we should add this to our style guide if we're going to enforce it, as it's not part of Google JS style.

gauntface commented on infra/testing/sw-env.js:
```diff
@@ -11,18 +11,34 @@
-global.fetch = mockFetch;
+// Assign all properties of `self` to `global`;
+Object.assign(global, serviceWorkerMock());
+global.self = global;
```
I'm still confused at why this self is needed. serviceWorkerMock creates a self and this will overwrite that.

philipwalton replied:
It's not necessarily needed, but if we don't do this then all the code below would need to be like this:

```js
global.Blob = global.self.Blob = Blob;
global.Event = global.self.Event = Event;
// ...
```

So I guess it's a matter of which we prefer. I think what I have is fine since it's actually the same way browsers implement it, i.e. self (and window) is a property of the global object that points to the global object. For example: `window.window === window`, `window.self === window`, `self.self === self`, etc.

gauntface commented on packages/workbox-background-sync/lib/Queue.mjs:
```diff
@@ -0,0 +1,208 @@
+   *     in IndexedDB specific to this instance. An error will be thrown if
+   *     a duplicate name is detected.
+   * @param {Object} [param2]
+   * @param {Object} [param2.callbacks] Callbacks to observe the lifecycle of
```
Is it fair to say that these callbacks are the same as plugins? If so can we rename this to plugins rather than callback for consistency throughout workbox.

philipwalton replied:
I don't think they're the same. As far as I'm aware, a plugin in an object that implements lifecycle callbacks, and this doesn't do that.

I've debated with myself as to whether or not these should be events and the `Queue` class inherit from something like `EventEmitter`, but I didn't want to do that in this PR to keep things as close to the original as possible.

gauntface commented on packages/workbox-background-sync/lib/Queue.mjs:
```diff
@@ -0,0 +1,208 @@
+   * @param {Object} [param2.callbacks] Callbacks to observe the lifecycle of
+   *     queued requests. Use these to respond to or modify the requests
+   *     during the replay process.
+   * @param {function(StorableRequest):undefined}
```
What is the syntax for `:undefined`? The parameter is marked as optional given the square brackets.

philipwalton replied:
This means the function returns `undefined`. The syntax is described here: https://github.com/google/closure-compiler/wiki/Types-in-the-Closure-Type-System

But that brings up a good point that this is closure compiler jsdoc notation, and I'm not sure if what we use to generate docs will support this. I don't think non-CC-jsdoc syntax has a way to specify what params a function takes, what its `this` binding is, or what it returns (which is too bad, IMO).
